### PR TITLE
DR | Refactor date error messages

### PR DESCRIPTION
**Pull request by rocio-desantiago**
## Summary

-   Renamed parameters, functions, object keys and variables (i.e,   `blankDecisionDate`, `rawDateString`)
-   Moved a lot of shared date functions from each form's date validations file into a shared date validations file. (i.e, `createScreenReaderErrorMsg`, `createDateObject`, `addDateErrorMessages`

## Why is this the solution?

- Renaming the variables enables for better code readability
- This reduces repetitive code across the 3 forms and enables for all 3 forms to be updated from one file and makes it easier to manage changes so that one form doesn't become outdated.

## Team
- Benefits Team 1

## Related issue(s)

- _Link to ticket created in va.gov-team repo_
department-of-veterans-affairs/va.gov-team#63177

## Testing done

- _Describe what the old behavior was prior to the change_
  > Unit tests
- _Describe the steps required to verify your changes are working as expected_
  > N/A
- _Describe the tests completed and the results_
  > tests not affected - no changes

## Screenshots

No UI changes

## What areas of the site does it impact?\

Appeals

*(Describe what parts of the site are impacted **if** code touched other areas)*

None

## Acceptance criteria

- [ ] Code standards reviewed by fellow developers on team.

### Quality Assurance & Testing

- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### CODE REVIEW

File: src/applications/appeals/shared/validations/date.js
Comment (Mottie) [inline src/applications/appeals/shared/validations/date.js:57]: Please replace the `invalidDate` with `dateString`. `invalidateDate` is a boolean. This looks like it might be a bug in the original code.

File: src/applications/appeals/995/validations/date.js
Comment (Mottie) [inline src/applications/appeals/995/validations/date.js:24]: I think we can replace all of these comparsons with `dateErrorMsgs` once we figure out which error messages to use.

File: src/applications/appeals/shared/validations/date.js
Comment (Mottie) [reply src/applications/appeals/shared/validations/date.js:57]: It might be a good idea to add `PropTypes` for all these functions.

**Review by Mottie — CHANGES_REQUESTED** [review]

File: src/applications/appeals/shared/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:57]: With it not being a component, we can't add PropTypes since these are functions.

File: src/applications/appeals/995/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/995/validations/date.js:24]: Yes, once content has been updated by design, we'll be able to consolidate the error messages to be shared across all 3

**Review by Mottie — APPROVED** [review]

File: src/applications/appeals/shared/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:16]: Copied the same code comment from other date files.

File: src/applications/appeals/shared/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:42]: Will not refactor `dateErrorMssgs` because of a slight difference in 995's date file.
>    } else if (hasErrorDate) {
>     errors.addError(errorMessages.invalidDate);

File: src/applications/appeals/995/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/995/validations/date.js:16]: `rawDateString` updated but holding off on `validateDate` because it looks like it'll be a more invasive change as it's called 61 times across 13 files.

File: src/applications/appeals/shared/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:35]: Reordered `isNaN(year)` ✅   
Robin mentioned we won't need to check if `year === '0'` because minYear is `1900`

File: src/applications/appeals/shared/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:43]: Possibly:
 - `invalidDate` to `isDateTooShortOrInvalid` or `isDateInvalidOrTooShort`
 - `hasErrorParts` to `hasDateErrors` would be more descriptive

I'm still trying to understand isValid through moment's [docs](https://momentjscom.readthedocs.io/en/latest/moment/01-parsing/15-is-valid/#momentisvalid).

I do remember we tried to take out `dateString?.length < FORMAT_YMD.length` from one of these two: `validateDateString` and `invalidDate` but tests broke where React couldn't even tell me why it was breaking.

File: src/applications/appeals/shared/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:35]: Adding comment ^

File: src/applications/appeals/shared/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:43]: Rename `missingDate` to `blankDate`

Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:43]: Oooh `invalidDate` exists 50 times across 18 files. Maybe also hold off on that?

File: src/applications/appeals/shared/validations/date.js
Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:64]: Considered moving `datePartErrors` but that one would require passing `month, day, year, maxDays`

even `todayOrFutureDate` would still require passing `date` in.

Comment (rocio-desantiago) [reply src/applications/appeals/shared/validations/date.js:64]: `dateErrorMsgs` renamed to `addDateErrorMessages`

File: src/applications/appeals/shared/validations/date.js
Comment (data-doge) [inline src/applications/appeals/shared/validations/date.js:25]: Let's stick with `isInvalidDateString`. `isDateStringValid` implies that the string is valid if this function returns `true`, which isn't the case.

File: src/applications/appeals/10182/validations/date.js
Comment (data-doge) [inline src/applications/appeals/10182/validations/date.js:63]: Where in the code does this come into play?

Comment (rocio-desantiago) [reply src/applications/appeals/10182/validations/date.js]: This creates the error messages to be read out by the screenreader.

**Review by data-doge — APPROVED** [review]
Excellent work! Let's wait until Monday to merge.

**Pull request merged**
